crates/ra_hir/src/ty.rs

```diff
@@ -872,6 +873,133 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
         }
     }
 
+    fn resolve_fields(&self, path: Option<&Path>) -> Option<(Ty, Vec<StructField>)> {
+        let def_id = self.module.resolve_path(self.db, path?).take_types()?;
+        let def = def_id.resolve(self.db);
+
+        match def {
+            Def::Struct(s) => {
+                let fields: Vec<_> = self
+                    .db
+                    .struct_data(s.def_id())
+                    .variant_data
+                    .fields()
+                    .to_owned();
+                Some((type_for_struct(self.db, s), fields))
+            }
+            Def::EnumVariant(ev) => {
+                let fields: Vec<_> = ev.variant_data(self.db).fields().to_owned();
+                Some((type_for_enum_variant(self.db, ev), fields))
+            }
+            _ => None,
+        }
+    }
+
+    fn infer_tuple_struct_pat(&mut self, path: Option<&Path>, subpats: &[PatId]) -> Ty {
+        let (ty, fields) = self
+            .resolve_fields(path)
+            .unwrap_or((Ty::Unknown, Vec::new()));
+
+        for (i, &subpat) in subpats.iter().enumerate() {
+            let expected_ty = fields
+                .get(i)
+                .map_or(Ty::Unknown, |field| self.make_ty(&field.type_ref));
```

Comment:
I just realized, this is wrong - it's resolving the field's type reference in the scope of the function that's being inferred. We already have a `StructField::ty` function that should instead be used.